### ISSUE TITLE
When disk full errors cause operation to terminate, VFS cache doesn't clear

### ISSUE THREAD

**Issue by dr-mcgillicuddy** [rclone/rclone#4442]
#### What is the problem you are having with rclone?
If VFS cache grows to 100% of the free space on a drive, rclone does not purge the cache on unmount, shutdown, reboot, or when running `--cache-db-purge`

Solution seems to be `rm -r ~/.cache/rclone/vfs/*` if one can get there without a system lockup.

In my specific case, I have a google drive mounted (with crypt) to a local directory, and in this instance was running wget on approximately 15,000 files totaling 30GB.  Wget downloads to the local rclone mount, but as my connection is asymmetrical, VFS filled up faster than it could be pushed up to Google, a bus error results, and the VFS cache must be manually cleared to restore the system to a usable state.

#### What is your rclone version (output from `rclone version`)
rclone v1.52.2
- os/arch: linux/amd64
- go version: go1.14.4

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Ubuntu 18.04, 64 bit

####  Which cloud storage system are you using? (eg Google Drive)

Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone mount crypt: ~/rclone --vfs-cache-mode full`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
Unavailable, and I no longer have the bandwidth to re-create this month.  If necessary, I can update the issue when my bandwidth counter resets.

**Comment by ncw** [comment]
Am I correct in thinking that the cache was full of files that rclone needed to upload? Rclone won't delete these automatically as they may contain data you need.

It sounds like rclone's handling of the disk full error could be better though!

Maybe rclone should monitor the total disk available in the cache directory. Not sure how reliable that is though.

**Comment by dr-mcgillicuddy** [comment]
`Am I correct in thinking that the cache was full of files that rclone needed to upload?`
I think you are correct.  However, there's no way to resume the upload to clear the cache.  I considered that simply re-mounting the drive might allow the uploads to complete.  However, FUSE won't work when the disk is 100% full (I forget the particular error it throws, though).  So the only thing I could come up with was just deleting the VFS cache.

I also attempted to run `rclone copy` on the cache directory itself, but that also failed due to lack of space on the local drive required to start the operation.

**Comment by ncw** [comment]
What do you think rclone should do here? It can be quite tricky to recover from disk full.

**Comment by dr-mcgillicuddy** [comment]
Oof. Disk full is a serious error with lots of complications, so that's a weighty decision to make.

In my personal case, I would have been satisfied if rclone had either:

1. Given an out of space error _before_ the drive was 100% full, so that the system was more functional, or
2. Responded to --cache-db-purge by wiping the VFS cache

The problem I see from rclone's perspective, is that whatever is filling up the VFS is likely not something rclone itself can control.  In my case it was wget; and rclone, obviously, has no means of pausing, slowing, or stopping/resuming wget.

All rclone could do, I think, is throw an out of space error before the drive was truly out of space.  This would force anything writing to the FUSE mount to terminate, and leave it up to the developers of that application to come up with a clean way of doing exiting on an out of space error.

`--vfs-cache-max-size` doesn't _quite_ do that, so maybe that flag in combination with `--rclone-mount-pseudo-size` or something, such that any non-rclone processes operating on a FUSE mount see a reasonable size storage bucket, rather than an essentially infinite one.  MacOS always reports my rclone mounts as:
Size: 17.59 TB
Available:17.59TB
Used: 0 Bytes

Maybe `--report-space-remaining`?
If FUSE reports "available space" in an amount less than the capacity of the drive where the VFS cache is held, would that break anything?  Anything that needs to pre-allocate space might get confused.  Also, sometimes the whole point of using rclone is that I _want_ my system to see an infinite storage bucket (e.g. for full system deduped backups).

I guess another way to put that is "What if --vfs-cache-max-size had a default value of something other than "off" and FUSE reported the available space on the mount point as:
Available Space = (VFS cache max size) - (VFS cache current size)?

My gut says this may stand to break more than it fixes.

If all of that carries too much risk of breaking something else, then maybe just a flag to flush the VFS, upload status notwithstanding, would do the trick in terms of salvaging stability of the system: `--vfs-force-empty`
or
`--vfs-upload-only` to have rclone start, upload the contents of vfs, purge the local directory, and then exit?

I feel wholly unqualified to opine on this, because while allowing anything to use 100% of disk space and then hard exit is problematic from a data integrity standpoint, I fear changing _something_ about that behavior may break _everything_ for someone.

**Comment by animosity22** [comment]
Hmm, I think maybe something along the lines of the upload only part so you can run a command to clear out the cache before starting it.

I think the goal would be to make sure data is not lost so the idea of that recovery mode is more on what I was thinking as well.

Having a full disk generally makes things go boom so recovering is always tricky.

**Comment by dr-mcgillicuddy** [comment]
> Hmm, I think maybe something along the lines of the upload only part so you can run a command to clear out the cache before starting it.

That's probably I feature I would use on a regular basis, even if I hadn't encountered a disk-full situation.
As it is, there's not really a way to tell if the VFS has fully uploaded to your storage bucket.

When I'm done using rclone, I sometimes use bmon to check upload bandwidth usage (if ~0 then I assume everything has synced).  Other times I let the system sit for a few hours, just for good measure.

If there was an "upload only and then quit" flag, one could:

`fuser mount -u ~/rclonemount`
then
`rclone --vfs-upload-only -P`

This, or something like it, would serve the dual purpose of providing the user with feedback on whether all queued VFS operations had completed, AND of clearing out disk space in the out-of-memory scenario I initially described.

If the disk is truly 100% full, however, rclone will not load at all.  There would need to be some sort of safeguard against 100% usage (say, soft-exit at some degree of full, either as a percentage or a nominal value), or such an upload command wouldn't work.

**Comment by ncw** [comment]
We merged into 1.53 code to clean the cache on disk full errors.

I think this should make it so this issue does not occur, or make it much less likely

**Comment by ferferga** [comment]
@ncw Hello and happy new year beforehand:

I think this issue it's not fully solved. I have my cache in ``/dev/shm`` in order to no burn my Pi's SD with a max size of 1G. When memory pressure increases and reclaiming space from that area is needed, the kernel removes files there and that's handled OK by rclone, which redownloads the chunks. However, I had troubles accessing my files today and this is what I saw in the logs:

```
ERROR : /folder/file.extension: vfs cache: failed to open item: vfs cache item: create cache file failed: vfs cache item: _save failed: vfs cache item: failed to encode metadata: write /dev/shm/rclone/AliasRemote/vfsMeta/CryptedRemote/folder/file.extension: no space left on device.
```

I have my mount as read only, which means that there's nothing useful in the cache that can't be redownloaded by rclone again. Here's my full mount command:
```
rclone mount --read-only \
                                --allow-other \
                                --use-mmap \
                                --cache-dir /dev/shm/rclone/AliasRemote \
                                --vfs-cache-mode full \
                                --vfs-cache-max-size=1G \
                                --vfs-read-chunk-size=128M \
                                --vfs-read-ahead=128M \
                                --buffer-size=32M \
                                AliasRemote: /mountpoint
```

This is my remote layout:

``Google Drive (Remote) > Crypt (CryptedRemote) > Alias (Alias remote, which wraps CryptedRemote:Folder1/Folder2)``

**rclone version output**:
```
rclone v1.57.0
- os/version: debian 11.2
- os/kernel: 5.10.63-v7l+ (armv7l)
- os/type: linux
- os/arch: arm
- go/version: go1.17.2
- go/linking: static
- go/tags: none
```

**Comment by animosity22** [comment]
It's best to open a forum post and share all the details in the help template..

**Comment by darthShadow** [comment]
Closing since the original issue has been resolved.
